### PR TITLE
RavenDB-18852 RangeQueries integration

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18852 

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+                //(x, y>
+                (UnaryMatchOperation.GreaterThan, UnaryMatchOperation.LessThanOrEqual) => RangeBuilder<TScoreFunction, Range.Exclusive, Range.Inclusive>(names.FieldName,
+                    names.NumericTree, (long)(object)low, (long)(object)high, scoreFunction, isNegated),
+            };
+        }
+
+        if (typeof(TValue) == typeof(double))
+        {
+            return (leftSide, rightSide) switch
```
Comment (arekpalinski) [inline src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs:43]: Can you please check the warning?

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public MultiTermMatch BetweenQuery<TValue, TScoreFunction>(string field, TValue low, TValue high,
+        TScoreFunction scoreFunction = default, UnaryMatchOperation leftSide = UnaryMatchOperation.GreaterThanOrEqual, UnaryMatchOperation rightSide = UnaryMatchOperation.LessThanOrEqual, bool isNegated = false,
+        int fieldId = Constants.IndexSearcher.NonAnalyzer)
+        where TScoreFunction : IQueryScoreFunction
+    {
+        var names = GetSliceForRangeQueries(field, low);
+        if (typeof(TValue) == typeof(long))
+        {
+            return (leftSide, rightSide) switch
```
Comment (arekpalinski) [inline src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs:20]: Can you please check the warning?

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+                    names.NumericTree, (double)(object)low, (double)(object)high, scoreFunction, isNegated),
+            };
+        }
+
+        if (typeof(TValue) == typeof(string))
+        {
+            var leftValue = EncodeAndApplyAnalyzer((string)(object)low, fieldId);
+            var rightValue = EncodeAndApplyAnalyzer((string)(object)high, fieldId);
+
+            return (leftSide, rightSide) switch
```
Comment (arekpalinski) [inline src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs:70]: And here as well

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                //(x, y>
+                (UnaryMatchOperation.GreaterThan, UnaryMatchOperation.LessThanOrEqual) => RangeBuilder<TScoreFunction, Range.Exclusive, Range.Inclusive>(names.FieldName,
+                    leftValue, rightValue, scoreFunction, isNegated),
+            };
+        }
+
+        throw new ArgumentException($"{typeof(TValue)} is not supported in {nameof(BetweenQuery)}");
+    }
+    //
+    // [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (arekpalinski) [inline src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs]: Commented code

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs [hunk 5]
```diff
@@ -169,78 +180,176 @@ public bool TryAnd(CoraxBooleanItem item)
...
+                _ => throw new InvalidOperationException($"UnaryMatchOperation {leftmostClause.Operation} is not supported for type {leftmostClause.Term.GetType()}")
+            };
+
+            baseMatch = baseMatch is null
+                ? nextQuery
+                : _indexSearcher.And(baseMatch, nextQuery);
+        }
+        else
+        {
+            var nextQuery = (leftmostClause.Operation, leftmostClause.Term) switch
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs:283]: Another warning

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs [hunk 6]
```diff
@@ -259,3 +368,52 @@ public IQueryMatch Materialize()
...
+            else if (item.Operation == UnaryMatchOperation.NotEquals)
+                ratio = (double)(_numberOfEntries - item.Count) / _numberOfEntries;
+            else
+            {
+                //This is ideal scenario IF they are ranged (not (-inf, X) or (X, inf)
+                //Values over 1 means duplications
+                var count = item.Materialize().Count;
+            }
+
+            //This is interesting part because we don't know how many items do we hit. Maybe should we go trough tree and calculate the cost first? Having
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs]: @ayende @redknightlois Any advice on that?

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs:20]: done

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs:43]: done

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs (hunk 3 above)
Comment (maciejaszyk) [reply src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Ranges.cs:70]: done

**Comment by arekpalinski** [comment]
@ayende @redknightlois can you please review it?

File: src/Corax/Queries/MultiUnaryMatch.cs [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+        }
+        
+        return result;
+    }
+
+    public void Score(Span<long> matches, Span<float> scores)
+    {
+        //TODO Federico: Can you handle that thing? I'm not sure how internally boosting flow works.
+        throw new NotImplementedException();
+    }
```
Comment (redknightlois) [inline src/Corax/Queries/MultiUnaryMatch.cs:554]: @maciejaszyk Totally forgot about this. Let's setup a call for tomorrow and we can write the code in there.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs [hunk 8]
```diff
@@ -649,15 +649,24 @@ CoraxBooleanItem HandleStringBetween()
...
-        return rawQuery.IsBoosting
-            ? rawQuery
-            : indexSearcher.Boost(rawQuery, scoreFunction);
+
+        if (rawQuery is CoraxBooleanItem cbi)
+            rawQuery = cbi.Materialize();
+        else if (rawQuery is CoraxBooleanQuery cbq)
+        {
+            rawQuery = cbq.Materialize();
+            //todo recieve hasBinary
```
Comment (redknightlois) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs]: Is this TODO still valid?

File: src/Corax/Queries/MultiUnaryMatch.cs (hunk 7 above)
Comment (maciejaszyk) [reply src/Corax/Queries/MultiUnaryMatch.cs:554]: done

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs (hunk 8 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs]: done

**Pull request merged**